### PR TITLE
Record iterator

### PR DESCRIPTION
**Pull request by aswan**

### CODE REVIEW

File: zng/recorditer.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	"github.com/brimsec/zq/zcode"
+)
+
+var (
+	ErrExhausted = errors.New("called Next() on iterator after last record")
+	ErrMismatch  = errors.New("mismatch between record type and value")
+)
+
+type RecordIter interface {
```
Comment (nwt) [inline zng/recorditer.go:15]: Why make this an interface?

Comment (aswan) [reply zng/recorditer.go]: no good reason, left over from early hacking.  removed.

File: zng/recorditer.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	stack []iterInfo
+}
+
+func (r *Record) NewFieldIterator() recordIter {
+	return recordIter{
+		stack: []iterInfo{iterInfo{
+			iter:     r.ZvalIter(),
+			typ:      r.Type,
+			offset:   0,
+			fullname: "",
```
Comment (nwt) [inline zng/recorditer.go:32]: Nit: Our prevailing convention is simply to omit zero fields.

File: zng/recorditer.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	typ      *TypeRecord
+	offset   int
+	fullname string
+}
+
+type recordIter struct {
+	stack []iterInfo
+}
+
+func (r *Record) NewFieldIterator() recordIter {
```
Comment (nwt) [inline zng/recorditer.go:26]: Nits: We usually a type's methods in the same file, so put this in recordval.go. And call it call it FieldIter so it agrees with (*Record).ZvalIter.

File: zng/recordval.go [hunk 4]
```diff
@@ -107,8 +107,14 @@ func (r *Record) ZvalIter() zcode.Iter {
 	return r.Raw.Iter()
 }
 
-// Width returns the number of columns in the record.
-func (r *Record) Width() int { return len(r.Type.Columns) }
+func (r *Record) NewFieldIter() recordIter {
```
Comment (nwt) [inline zng/recordval.go:110]:
```suggestion
func (r *Record) FieldIter() recordIter {
```

File: zng/recorditer.go (hunk 3 above)
Comment (henridf) [reply zng/recorditer.go]: nit: should recordIter be called fieldIter (or vice versa)?

File: zng/recorditer_test.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+	name, val, err = it.Next()
+	require.NoError(t, err)
+	require.Equal(t, "r1.a", name, "got correct field name")
+	l, err := val.ContainerLength()
+	require.NoError(t, err)
+	require.Equal(t, 3, l, "got array of length 3")
+	require.False(t, it.Done(), "iterator is not exhausted")
+
+	name, val, err = it.Next()
```
Comment (henridf) [inline zng/recorditer_test.go:46]:
```suggestion
	name, val, err = it.Next()
	require.NoError(t, err)
```

**Comment by henridf** [comment]
I'm guessing the @mccanne's indexer could use this...

**Comment by aswan** [comment]
> I'm guessing @mccanne's indexer could use this...

probably.  it needs somebody to approve it first though

**Comment by nwt** [comment]
It looks like you've rebased a bunch of commits from master onto this, making the diff hard to follow.

**Comment by aswan** [comment]
> It looks like you've rebased a bunch of commits from master onto this, making the diff hard to follow.

I meant to merge but I guess I botched it, do you want me to clean it up and force-push so you can take anotehr pass?

**Comment by nwt** [comment]
It'd just be nice if the PR were readable, for posterity. But if it's hard for you to fix, don't sweat it.

**Comment by henridf** [comment]
just noticed recorditer.go should be changed to fielditer.go (and similarly for test file...)
